File: 19-remove-nth-node-from-end-of-list/python3/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head == None:
            return None
        st = []
        while head:
            st.append(head)
            head = head.next
        tail = None
        while n > 1:
            tail = st.pop()
            n -= 1
        nth = st.pop()
        nth.next = None
        if st:
            st[-1].next = tail
        return st[0] if st else tail

File: 19-remove-nth-node-from-end-of-list/java/19-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode removeNthFromEnd(ListNode head, int n) {
    var queue = new ArrayDeque<ListNode>();
    var res = head;
    while (null != head) {
      while (queue.size() > n) {
        queue.pollFirst();
      }

      queue.add(head);
      head = head.next;
    }

    if (queue.size() == n) {
      queue.pollFirst();
      res = queue.peekFirst();
    } else {
      var node = queue.pollFirst();
      var to_delete = queue.pollFirst();
      node.next = to_delete.next;
    }

    return res;
  }
}
